### PR TITLE
Combo: Panel Cp (tandem-only, ×0.1) + AoA Curriculum (warmup=20)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Our two most promising experiments this round target DIFFERENT metrics through DIFFERENT mechanisms:

1. **Panel-Method Cp Feature (tandem-only)** — p_oodc -2.8%, p_tan -2.9% (thorfinn #2319 v2). Physics-informed input feature that helps tandem/OOD predictions.
2. **AoA Curriculum Training (warmup=40)** — p_oodc -5.1% (fern #2328 v1). Training procedure that improves OOD generalization through structured sample exposure.

These are **orthogonal improvements** — one is an input feature, the other is a training schedule. They should stack. The combination could achieve p_oodc improvements of -5% to -8% while potentially mitigating each other's weaknesses:
- The panel Cp feature's p_in regression may be reduced by the curriculum's p_in improvement (-0.5%)
- The curriculum's p_tan regression may be reduced by the panel Cp's p_tan improvement (-2.9%)

**This is a bold combination test.** We implement BOTH changes simultaneously to see if the improvements compound.

## Instructions

You need to implement TWO features from the thorfinn and fern branches. Look at their PRs for implementation reference:
- **PR #2319** (thorfinn): Panel Cp computation + --cp_panel + --cp_panel_tandem_only flags
- **PR #2328** (fern): AoA curriculum + --aoa_curriculum + --aoa_curriculum_warmup flags

### Step 1: Cherry-pick or re-implement both features

Check out the thorfinn and fern branches to study their implementations:
```bash
git diff noam..origin/thorfinn/panel-method-cp-feature -- cfd_tandemfoil/train.py
git diff noam..origin/fern/aoa-curriculum-training -- cfd_tandemfoil/train.py
```

Re-implement both features in your branch. Key components:
1. **Panel Cp feature:** compute_cp_panel() function, --cp_panel flag, --cp_panel_tandem_only flag, feature concatenation
2. **AoA curriculum:** get_curriculum_weights() function, --aoa_curriculum flag, --aoa_curriculum_warmup flag, per-epoch sampler weights

### Step 2: Add the combination + scaling

Apply panel Cp with 0.1 scaling for tandem samples only:
```python
if args.cp_panel and args.cp_panel_tandem_only:
    if is_tandem:
        cp_feat = cp_feat * 0.1  # Scale down
    else:
        cp_feat = torch.zeros_like(cp_feat)
```

### Step 3: Run 2 seeds

```bash
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/combo-cp-curriculum-s42" --seed 42 \
  --wandb_group panel-cp-plus-curriculum \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --cp_panel --cp_panel_tandem_only \
  --aoa_curriculum --aoa_curriculum_warmup 20

# Repeat for seed 73: --seed 73 --wandb_name "askeladd/combo-cp-curriculum-s73"
```

**Key flags added:** --cp_panel --cp_panel_tandem_only --aoa_curriculum --aoa_curriculum_warmup 20

## Baseline

| Metric | Baseline | Panel Cp v2 | AoA Curriculum v1 | Target (combo) |
|--------|----------|------------|-------------------|----------------|
| p_in | 11.742 | 12.265 (+4.5%) | 11.687 (-0.5%) | < 11.742 |
| p_oodc | 7.643 | 7.426 (-2.8%) | 7.252 (-5.1%) | < 7.2 (stacked) |
| p_tan | 27.874 | 27.063 (-2.9%) | 28.859 (+3.5%) | < 27.874 |
| p_re | 6.419 | 6.432 (+0.2%) | 6.500 (+1.3%) | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Combo: Panel Cp (tandem-only, ×0.1) + AoA Curriculum (warmup=20)

| Metric | Baseline | S42 | S73 | 2-seed Avg | Δ vs Baseline |
|--------|----------|-----|-----|------------|--------------|
| **p_in** | 11.742 | 11.732 | 12.395 | **12.064** | **+2.7%** ❌ |
| **p_oodc** | 7.643 | 7.719 | 7.525 | **7.622** | **-0.3%** (flat) |
| **p_tan** | 27.874 | 28.935 | 28.039 | **28.487** | **+2.2%** ❌ |
| **p_re** | 6.419 | 6.641 | 6.545 | **6.593** | **+2.7%** ❌ |
| val/loss | ~0.37 | 0.3792 | 0.3764 | 0.3778 | ~ baseline |

### Commands

```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/combo-cp-curriculum-s42" --seed 42 \
  --wandb_group panel-cp-plus-curriculum \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --cp_panel --cp_panel_tandem_only \
  --aoa_curriculum --aoa_curriculum_warmup 20
# Seed 73: --seed 73 --wandb_name "askeladd/combo-cp-curriculum-s73"
```

### Details

- **Peak memory:** 46.6 GB (S42), 46.3 GB (S73)
- **Epochs:** 145 (both hit 180-min wall-clock limit)
- **W&B runs:** [0vv815dd](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/0vv815dd) (S42), [xn59kd8e](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xn59kd8e) (S73)

### What happened

**The combination did NOT stack as hoped.** The two features interfered rather than complementing each other:

- **p_oodc** — essentially flat (-0.3%), far from the hoped -5% to -8% stacking. The individual improvements (Panel Cp -2.8%, AoA curriculum -5.1%) cancelled each other out when combined.
- **p_in** — regressed +2.7%, worse than either feature alone. Panel Cp's p_in regression (+4.5%) dominated and the curriculum's slight p_in benefit (-0.5%) couldn't compensate.
- **p_tan** — regressed +2.2%, worse than baseline despite Panel Cp alone giving -2.9%. The curriculum's p_tan regression (+3.5%) swamped the Cp benefit.
- **p_re** — regressed +2.7% across both seeds. Neither feature individually helped Re generalization, and the combination made it worse.

The val/loss is ~baseline (0.378 vs ~0.37), confirming the features don't break training but don't improve the loss landscape when combined.

**Key insight:** The Panel Cp feature and AoA Curriculum target overlapping aspects of the model's learning — both influence how the model handles angle-of-attack variation. When combined, they likely introduce conflicting gradients: the curriculum wants the model to focus on easy (low-AoA) samples early while the Cp feature is most informative at high AoA. The 0.1 scaling on the Cp feature may also have been too aggressive, effectively turning it into noise during the curriculum warmup phase.

### Suggested follow-ups

1. **Try them separately with more epochs.** Both features showed promise individually with fewer epochs. The 145 epochs here should have been ample for convergence — the issue is the combination, not the training duration.
2. **If combining, try sequential:** first train with AoA curriculum alone, then fine-tune with Cp feature added (or vice versa), rather than simultaneous training.
3. **Adjust Cp scaling.** The 0.1 scale factor may be too small — try 0.3 or 0.5 to give the Cp feature more signal.
4. **Delayed Cp injection:** Only add the Cp feature after the curriculum warmup completes (epoch > warmup), so the two mechanisms don't conflict during early training.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2336 because Panel Cp + AoA Curriculum combo does not stack: p_in +3.0%, p_oodc +1.0%, p_tan +4.0%, p_re +1.7%. The two approaches target overlapping aspects (AoA handling) and produce conflicting gradients. Good analysis from askeladd — the curriculum warmup deprives Cp features of signal during early training. AoA curriculum is now exhausted in all forms (standalone, warmup variations, combinations). W&B: 0vv815dd (s42), xn59kd8e (s73)..
